validphys2/src/validphys/photon/compute.py

```diff
@@ -0,0 +1,265 @@
+        return self.alpha_em_nlo(
+            q,
+            self.alpha_thresh[nf],
+            self.thresh[nf],
+            nf
+        )
+    
+    def alpha_em_nlo(self, q, alpha_ref, qref, nf):
+        """
+        Compute the alpha_em running for FFS at NLO.
+
+        Parameters
+        ----------
+        q : float
+            target scale
+        a_ref : float
+            reference value of a = alpha_em/(4*pi)
+        qref: float
+            reference scale
+        nf: int
+            number of flavors
+
+        Returns
+        -------
+        as_NLO : float
+            target value of a
+        """
+        lmu = 2 * np.log(q / qref)
+        den = 1.0 + self.beta0[nf] * alpha_ref * lmu
+        alpha_LO = alpha_ref / den
+        alpha_NLO = alpha_LO * (1 - self.b1[nf] * alpha_LO * np.log(den))
+        return alpha_NLO
+    
+    def set_thresholds_alpha_em(self):
+        """Compute and store the couplings at thresholds"""
+        thresh_list = [self.thresh_c, self.thresh_b, self.thresh_t]
+        if self.qref < self.thresh_c :
+            nfref = 3
+        elif self.qref < self.thresh_b :
+            nfref = 4
+        elif self.qref < self.thresh_t :
+            nfref = 5
+        else :
+            nfref = 6
+        thresh_list.insert(nfref - 3, self.qref)
+
+        self.thresh = {nf: thresh_list[nf - 3] for nf in range(3, self.theory["MaxNfAs"] + 1)}
+
+        self.alpha_thresh = {nfref: self.alpha_em_ref}
+
+        for nf in range(nfref + 1, self.theory["MaxNfAs"] + 1):
+            self.alpha_thresh[nf] = self.alpha_em_nlo(self.thresh[nf], self.alpha_thresh[nf - 1], self.thresh[nf - 1], nf - 1)
+        
+        for nf in reversed(range(3, nfref)):
+            self.alpha_thresh[nf] = self.alpha_em_nlo(self.thresh[nf], self.alpha_thresh[nf + 1], self.thresh[nf + 1], nf + 1)
+    
+    def set_betas(self):
+        """Compute and store beta0 / 4pi and b1 = (beta1/beta0)/4pi as a function of nf."""
+        nl = 3
+        nc = 3
+        eu2 = 4. / 9
+        ed2 = 1. / 9
+        self.beta0 = {}
+        self.b1 = {}
+        for nf in range(3, 6+1):
+            nu = nf // 2
+            nd = nf - nu
+            self.beta0[nf] = ( -4.0 / 3 * (nl + nc * (nu * eu2 + nd * ed2)) ) / (4 * np.pi)
+            self.b1[nf] = -4.0 * ( nl + nc * (nu * eu2**2 + nd * ed2**2) ) / self.beta0[nf] / (4 * np.pi)**2
+
+    def compute_photon_array(self, id):
+        r"""
+        Compute the photon PDF for every point in the grid xgrid.
+
+        Parameters
+        ----------
+        xgrids: numpy.array
+            grid of the x points
+        
+        Returns
+        -------
+        compute_photon_array: numpy.array
+            photon PDF at the scale 1 GeV
+        """
+        # Compute photon PDF
+        start_time = time.perf_counter()
+        photon_100GeV = np.array([self.lux[id].EvaluatePhoton(x, self.q_in2).total for x in self.xgrid])
+        photon_100GeV += self.generate_errors()
```

Comment:
I don't get why this is some random matrix instead of the usual "pdf covmat".

Reply:
This function implements the additional errors given by the LuxQED approach, i.e. it is the python rewriting of

https://github.com/NNPDF/nnpdf/blob/a91f38ed432723b23afa4b6d75345400d08cb95f/nnpdfcpp/src/fiatlux/src/fiatlux.cc#L75

(that is eq. 2.3 of the NNPDF3.1 paper)

Reply:
https://arxiv.org/pdf/1712.07053.pdf

Reply:
I'm sorry, it was the NNPDF3.1luxqed paper